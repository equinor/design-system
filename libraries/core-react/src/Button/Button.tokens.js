import primary from '@equinor/eds-tokens/components/button/buttons-primary.json'
import secondary from '@equinor/eds-tokens/components/button/buttons-secondary.json'
import danger from '@equinor/eds-tokens/components/button/buttons-danger.json'
import disabled from '@equinor/eds-tokens/components/button/buttons-disabled.json'

const colors = {
  primary: {
    ...primary,
    ghost_icon: {
      ...primary.ghost_icon,
      typography: primary.ghost.typography,
    },
  },
  secondary: {
    ...secondary,
    ghost_icon: {
      ...secondary.ghost_icon,
      typography: secondary.ghost.typography,
    },
  },
  danger: {
    ...danger,
    ghost_icon: {
      ...danger.ghost_icon,
      typography: danger.ghost.typography,
    },
  },
  disabled: {
    ...disabled,
    ghost_icon: {
      ...disabled.ghost_icon,
      typography: disabled.ghost.typography,
    },
  },
}

export const button = {
  colors,
}
