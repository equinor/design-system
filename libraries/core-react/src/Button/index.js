// @ts-nocheck
export { Button } from './Button'
