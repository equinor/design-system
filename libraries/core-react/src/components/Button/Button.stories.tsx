import { Button, Icon, ButtonProps, EdsProvider } from '../..'
import styled from 'styled-components'
import { Meta, Story } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { menu } from '@equinor/eds-icons'

const Wrapper = styled.div`
  margin: 32px;
  display: grid;
  grid-gap: 32px;
  grid-template-columns: repeat(4, fit-content(100%));
`

const FullWidthWrapper = styled.div`
  margin: 32px;
  display: grid;
  grid-gap: 16px;
`

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    as: {
      control: {
        type: 'select',
        options: ['span', 'a', 'button'],
        defaultValue: 'button',
      },
    },
  },
  parameters: {
    docs: {
      description: {
        component: `Buttons allow users to take action with a single click or tap.
        `,
      },
    },
    info: {},
  },
} as Meta

export const Default: Story<ButtonProps> = (args) => (
  <Button {...args}>You can control me</Button>
)

export const All: Story<ButtonProps> = () => (
  <Wrapper>
    <Button>Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
    <Button disabled>Disabled</Button>
    <Button variant="outlined">Primary</Button>
    <Button variant="outlined" color="secondary">
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      Danger
    </Button>
    <Button variant="outlined" disabled>
      Disabled
    </Button>
    <Button variant="ghost">Primary</Button>
    <Button variant="ghost" color="secondary">
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      Danger
    </Button>
    <Button variant="ghost" disabled>
      Disabled
    </Button>
    <Button variant="ghost_icon">
      <Icon name="save" title="save action"></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" disabled>
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Contained: Story<ButtonProps> = () => (
  <Wrapper>
    <Button>Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
    <Button disabled>Disabled</Button>
    <Button>
      <Icon name="save" title="save"></Icon>Primary
    </Button>
    <Button color="secondary">
      <Icon name="save" title="save"></Icon>Secondary
    </Button>
    <Button color="danger">
      <Icon name="save" title="save"></Icon>Danger
    </Button>
    <Button disabled>
      <Icon name="save" title="save"></Icon>Disabled
    </Button>
    <Button>
      Primary <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Outlined: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="outlined">Primary</Button>
    <Button variant="outlined" color="secondary">
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      Danger
    </Button>
    <Button variant="outlined" disabled>
      Disabled
    </Button>
    <Button variant="outlined">
      <Icon name="save" title="save"></Icon>
      Primary
    </Button>
    <Button variant="outlined" color="secondary">
      <Icon name="save" title="save"></Icon>
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      <Icon name="save" title="save"></Icon>
      Danger
    </Button>
    <Button variant="outlined" disabled>
      <Icon name="save" title="save"></Icon>
      Disabled
    </Button>
    <Button variant="outlined">
      Primary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="outlined" disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Ghost: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="ghost">Primary</Button>
    <Button variant="ghost" color="secondary">
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      Danger
    </Button>
    <Button variant="ghost" disabled>
      Disabled
    </Button>
    <Button variant="ghost">
      <Icon name="save" title="save"></Icon>
      Primary
    </Button>
    <Button variant="ghost" color="secondary">
      <Icon name="save" title="save"></Icon>
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      <Icon name="save" title="save"></Icon>
      Danger
    </Button>
    <Button variant="ghost" disabled>
      <Icon name="save" title="save"></Icon>
      Disabled
    </Button>
    <Button variant="ghost">
      Primary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" color="secondary">
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" color="danger">
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost" disabled>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const GhostIcon: Story<ButtonProps> = () => (
  <Wrapper>
    <Button variant="ghost_icon">
      <Icon name="save" title="save action"></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger">
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button variant="ghost_icon" disabled>
      <Icon name="save" title="save"></Icon>
    </Button>
  </Wrapper>
)

export const Form: Story<ButtonProps> = () => {
  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault() // to prevent navigation from storybook
    action('onSubmit')(e)
  }

  return (
    <Wrapper>
      <form onSubmit={handleSubmit}>
        <Button type="submit">Submit form</Button>
      </form>
    </Wrapper>
  )
}

export const FileUpload: Story<ButtonProps> = () => (
  <Wrapper>
    <input type="file" id="file-upload" style={{ display: 'none' }} multiple />
    <label htmlFor="file-upload">
      <Button as="span">Upload</Button>
    </label>
  </Wrapper>
)

FileUpload.parameters = {
  docs: {
    description: {
      story:
        'Please note this demo only works in Storybook Canvas (isolated example only)',
    },
  },
}

export const Link: Story<ButtonProps> = () => (
  <Wrapper>
    <Button href="#">Link</Button>
  </Wrapper>
)

export const DisabledLink: Story<ButtonProps> = () => {
  const isLinkValid = false
  return (
    <Wrapper>
      <Button
        href={isLinkValid ? '#' : undefined}
        disabled={isLinkValid ? false : true}
      >
        Disabled Link
      </Button>
    </Wrapper>
  )
}

DisabledLink.parameters = {
  docs: {
    description: {
      story:
        "Disabling a `Button` with a valid `href` URL does not work, as it's not possible to disable links. A workaround for this is to set `href={undefined}`.",
    },
  },
}

export const FullWidth: Story<ButtonProps> = () => (
  <FullWidthWrapper>
    <Button fullWidth>Primary</Button>
    <Button color="secondary" fullWidth>
      Secondary
    </Button>
    <Button color="danger" fullWidth>
      Danger
    </Button>
    <Button disabled fullWidth>
      Disabled
    </Button>
    <Button fullWidth>
      <Icon name="save" title="save"></Icon>Primary
    </Button>
    <Button color="secondary" fullWidth>
      <Icon name="save" title="save"></Icon>Secondary
    </Button>
    <Button color="danger" fullWidth>
      <Icon name="save" title="save"></Icon>Danger
    </Button>
    <Button disabled fullWidth>
      <Icon name="save" title="save"></Icon>Disabled
    </Button>
    <Button fullWidth>
      Primary <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="secondary" fullWidth>
      Secondary
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button color="danger" fullWidth>
      Danger
      <Icon name="save" title="save"></Icon>
    </Button>
    <Button disabled fullWidth>
      Disabled
      <Icon name="save" title="save"></Icon>
    </Button>
  </FullWidthWrapper>
)

export const Compact: Story<ButtonProps> = () => (
  <EdsProvider density="compact">
    <Wrapper>
      <Button>Contained</Button>
      <Button variant="outlined">Outlined</Button>
      <Button variant="ghost">Ghost</Button>
      <Button variant="ghost_icon">
        <Icon data={menu} title="Ghost icon menu"></Icon>
      </Button>
    </Wrapper>
  </EdsProvider>
)

Compact.parameters = {
  docs: {
    description: {
      story: 'Compact `Button` using `EdsProvider`',
    },
  },
}
