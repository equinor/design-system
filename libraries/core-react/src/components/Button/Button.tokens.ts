import { tokens } from '@equinor/eds-tokens'
import type {
  Border,
  Focus,
  Hover,
  Spacing,
  Typography,
} from '@equinor/eds-tokens'

const {
  components: {
    button: { primary, secondary, danger, disabled },
  },
} = tokens

const colors = {
  primary: {
    ...primary,
    ghost_icon: {
      ...primary.ghost_icon,
      typography: primary.ghost.typography,
    },
  },
  secondary: {
    ...secondary,
    ghost_icon: {
      ...secondary.ghost_icon,
      typography: secondary.ghost.typography,
    },
  },
  danger: {
    ...danger,
    ghost_icon: {
      ...danger.ghost_icon,
      typography: danger.ghost.typography,
    },
  },
  disabled: {
    ...disabled,
    ghost_icon: {
      ...disabled.ghost_icon,
      typography: disabled.ghost.typography,
    },
  },
}

export type Button = {
  height: string
  width?: string
  background: string
  color: string
  border: Border
  typography: Typography
  spacing?: Partial<Spacing>
  focus?: Focus
  hover?: Hover
  // TODO Remove these once figma-broker is updated with proper types
  pressedColor?: string
  clickboundOffset?: number | string
  clickbound?: string
}

type Buttons = {
  [P in keyof typeof colors]: {
    [P2 in keyof typeof colors[P]]: Button
  }
}

export type ButtonGroups =
  | Buttons['primary']
  | Buttons['secondary']
  | Buttons['danger']
  | Buttons['disabled']

export const button: {
  colors: Buttons
  icon_size: {
    width: string
    height: string
  }
} = {
  colors,
  icon_size: {
    width: '24px',
    height: '24px',
  },
}
