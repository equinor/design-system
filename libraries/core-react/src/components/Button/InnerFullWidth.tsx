import { forwardRef, Children as ReactChildren, HTMLAttributes } from 'react'
import styled from 'styled-components'

const FullWidthCenterContent = styled.span`
  text-align: center;
  flex: 1;
`

const FullWidthInner = styled.span`
  height: 100%;
  display: flex;
  align-items: center;

  > img:first-child,
  > svg:first-child {
    margin-right: 8px;
  }

  > img:last-child,
  > svg:last-child {
    margin-left: 8px;
  }

  > img:only-child,
  > svg:only-child {
    margin-left: auto;
    margin-right: auto;
  }

  > span:first-child {
    margin-left: 32px;
  }

  > span:last-child {
    margin-right: 32px;
  }

  > span:only-child {
    margin-right: 0;
    margin-left: 0;
  }
`

export const InnerFullWidth = forwardRef<
  HTMLSpanElement,
  HTMLAttributes<HTMLSpanElement>
>(function InnerFullWidth({ children }) {
  // We need everything in elements for proper flexing ðŸ’ª
  const updatedChildren = ReactChildren.map(children, (child) =>
    typeof child !== 'object' ? (
      <FullWidthCenterContent>{child}</FullWidthCenterContent>
    ) : (
      child
    ),
  )

  return <FullWidthInner>{updatedChildren}</FullWidthInner>
})
