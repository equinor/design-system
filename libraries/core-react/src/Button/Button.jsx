import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { Icon } from '..'
import { button } from './Button.tokens'
import { typographyTemplate } from '../_common/templates'

const { colors } = button
// TODO: Is there a better way to handle css properties without
// bloating with ${props => props.base.focus.color} etc...
const Base = ({ base, baseDisabled: disabled }) => {
  if (!base) {
    // TODO: What to do when base does not exist
    return ``
  }

  const { border, spacing, typography, focus, hover } = base

  return css`
    background: ${base.background};
    height: ${base.height};
    width: ${base.width};
    color: ${base.color};
    fill: ${base.color};
    svg {
      height: 16px;
      width: 16px;
    }

    border-radius: ${border.radius};
    border-color: ${border.color};
    border-width: ${border.width};

    ${spacing &&
      css`
        padding-left: ${spacing.left};
        padding-right: ${spacing.right};
      `}

    ${typographyTemplate(typography)}

    &::after {
      position: absolute;
      top: -${base.clickboundOffset};
      left: 0;
      width: 100%;
      height: ${base.clickbound};
      content: '';
    }

    &:hover {
      background: ${hover.background};
      ${hover.radius &&
        css`
          border-radius: ${hover.radius};
        `}
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      outline: ${focus.width} ${focus.type} ${focus.color};
      outline-offset: 2px;
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled {
      cursor: not-allowed;
      background: ${disabled.background};
      color: ${disabled.color};
      fill: ${disabled.color};

      border-color: ${disabled.border.color};

      &:hover {
        background: ${disabled.background};
      }
    }
  `
}

const ButtonBase = styled.button.attrs(({ type = 'button' }) => ({
  type,
}))`
  margin: 0;
  padding: 0;
  ${Base}
  position: relative;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  &::before {
    position: absolute;
    top: 0;
    left: 0;
    width: auto;
    min-height: auto;
    content: '';
  }
`

// TODO:
// - Missing top spacing
// - Fallback color?
// - Use ThemeProvider
// - Set attr on button

/**
 * @typedef Props
 * @prop {'contained' | 'outlined' | 'ghost' | 'ghost_icon'} [variant] Specifies which variant to use
 * @prop {boolean} [disabled] If `true`, the button will be disabled
 * @prop {'primary' | 'secondary' | 'danger'} [color] Specifies color
 */

/**
 * @param {Props & React.ButtonHTMLAttributes<HTMLButtonElement>} props
 */
export const Button = ({
  variant,
  children,
  disabled,
  className,
  color,
  ...other
}) => {
  const colorBase = colors[color] || {}
  const base = colorBase[variant] || {}
  const baseDisabled = colors.disabled[variant] || {}

  return (
    <ButtonBase
      base={base}
      baseDisabled={baseDisabled}
      className={className}
      disabled={disabled}
      {...other}
    >
      {children}
    </ButtonBase>
  )
}

Button.propTypes = {
  /** @ignore */
  children: ({ children }) => {
    let error

    const iconChildIsMissingTitle = React.Children.toArray(children)
      .map(
        // @ts-ignore
        ({ type, props: childProps }) =>
          // @ts-ignore
          (type || { displayName: '' }).displayName === Icon.displayName &&
          !childProps.title,
      )
      .includes(true)

    if (iconChildIsMissingTitle) {
      error = new Error(
        `When using an Icon in the Button, the property "title" is mandatory on Icon to meet accessibility requirements`,
      )
    }
    return error
  },
  /**  Specifies color */
  color: PropTypes.oneOf(['primary', 'secondary', 'danger']),
  /** Specifies which variant to use */
  variant: PropTypes.oneOf(['contained', 'outlined', 'ghost', 'ghost_icon']),

  /**
   * If `true`, the button will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * @ignore
   */
  className: PropTypes.string,
}

Button.defaultProps = {
  variant: 'contained',
  color: 'primary',
  disabled: false,
  className: '',
  children: null,
}

Button.displayName = 'eds-button'
