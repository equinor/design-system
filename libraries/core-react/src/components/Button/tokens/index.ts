import { ButtonTokenSet } from '../Button.types'

import * as contained from './contained'
import * as outlined from './outlined'
import * as ghost from './ghost'
import * as icon from './icon'

type ButtonTokens = {
  primary: ButtonTokenSet
  secondary: ButtonTokenSet
  danger: ButtonTokenSet
}

export const token: ButtonTokens = {
  primary: {
    contained: contained.primary,
    outlined: outlined.primary,
    ghost: ghost.primary,
    ghost_icon: icon.primary,
  },
  secondary: {
    contained: contained.secondary,
    outlined: outlined.secondary,
    ghost: ghost.secondary,
    ghost_icon: icon.secondary,
  },
  danger: {
    contained: contained.danger,
    outlined: outlined.danger,
    ghost: ghost.danger,
    ghost_icon: icon.danger,
  },
}
