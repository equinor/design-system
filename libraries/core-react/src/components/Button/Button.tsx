import { forwardRef, ElementType, ButtonHTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { token as buttonToken } from './tokens'
import { ButtonTokenSet, ButtonToken } from './Button.types'
import {
  typographyTemplate,
  bordersTemplate,
  outlineTemplate,
  spacingsTemplate,
} from '../../utils'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '../EdsProvider'

type Colors = 'primary' | 'secondary' | 'danger'
type Variants = 'contained' | 'outlined' | 'ghost' | 'ghost_icon'

const getVariant = (
  tokenSet: ButtonTokenSet,
  variant: Variants,
): ButtonToken => {
  switch (variant) {
    case 'ghost':
      return tokenSet.ghost
    case 'ghost_icon':
      return tokenSet.ghost_icon
    case 'outlined':
      return tokenSet.outlined
    case 'contained':
    default:
      return tokenSet.contained
  }
}

const getToken = (variant: Variants, color: Colors): ButtonToken => {
  switch (color) {
    case 'danger':
      return getVariant(buttonToken.danger, variant)
    case 'secondary':
      return getVariant(buttonToken.secondary, variant)
    case 'primary':
    default:
      return getVariant(buttonToken.primary, variant)
  }
}

const ButtonInner = styled.span`
  display: grid;
  grid-gap: 8px;
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;
`

const Base = ({ token, density }: { token: ButtonToken; density: string }) => {
  const { spacings, states } = token
  const { focus, hover, disabled } = states

  let height = token.height
  let width = token.width
  let clickbound = token.clickbound

  if (density === 'compact') {
    height = token.modes.compact.height
    width = token.modes.compact.width
    clickbound = token.modes.compact.clickbound
  }

  return css`
    background: ${token.background};
    height: ${height};
    width: ${width};
    svg {
      justify-self: center;
    }

    ${spacingsTemplate(spacings)}
    ${bordersTemplate(token.border)}
    ${typographyTemplate(token.typography)}

    &::after {
      position: absolute;
      top: -${clickbound?.offset?.top};
      left: -${clickbound?.offset?.left};
      width: ${clickbound?.width};
      height: ${clickbound?.height};
      content: '';
    }

    &:hover {
      background: ${hover.background};
      color: ${hover.typography?.color};
      ${bordersTemplate(hover?.border)}
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      ${outlineTemplate(focus.outline)}
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled {
      cursor: not-allowed;
      background: ${disabled.background};
      ${bordersTemplate(disabled.border)}
      ${typographyTemplate(disabled.typography)}

      &:hover {
        background: ${disabled.background};
      }
    }
  `
}

const ButtonBase = styled.button`
  margin: 0;
  padding: 0;
  ${Base}
  text-decoration: none;
  position: relative;
  cursor: pointer;
  display: inline-block;

  &::before {
    position: absolute;
    top: 0;
    left: 0;
    width: auto;
    min-height: auto;
    content: '';
  }
`
export type ButtonProps = {
  /**  Specifies color */
  color?: 'primary' | 'secondary' | 'danger'
  /** Specifies which variant to use */
  variant?: 'contained' | 'outlined' | 'ghost' | 'ghost_icon'
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Change html element. */
  as?: ElementType
  /** Type of button
   * @default 'button'
   */
  type?: string
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      ...other
    },
    ref,
  ) {
    const token = getToken(variant, color)
    const { density } = useEds()
    const as: ElementType =
      href && !disabled ? 'a' : other.as ? other.as : 'button'
    const type = href || other.as ? undefined : 'button'

    tabIndex = disabled ? -1 : tabIndex

    const buttonProps = {
      ref,
      as,
      href,
      type,
      token,
      disabled,
      tabIndex,
      density,
      ...other,
    }

    return (
      <ButtonBase {...buttonProps}>
        {fullWidth ? (
          <InnerFullWidth>{children}</InnerFullWidth>
        ) : (
          <ButtonInner>{children}</ButtonInner>
        )}
      </ButtonBase>
    )
  },
)
