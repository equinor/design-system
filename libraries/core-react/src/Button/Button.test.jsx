import React from 'react'
import Enzyme from 'enzyme'
import Button from './Button'

const { shallow, mount } = Enzyme

test('it matches snapshot', () => {
    const wrapper = shallow(<Button />)
    expect(wrapper).toMatchSnapshot()
})
