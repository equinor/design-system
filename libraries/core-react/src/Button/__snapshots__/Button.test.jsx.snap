// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Button-sc-17t57oa-0",
        "isStatic": false,
        "rules": Array [
          "background:",
          [Function],
          ";color:white;",
        ],
      },
      "defaultProps": Object {
        "variant": "primary",
      },
      "displayName": "eds-button",
      "foldedComponentIds": Array [],
      "propTypes": Object {
        "variant": [Function],
      },
      "render": [Function],
      "styledComponentId": "Button-sc-17t57oa-0",
      "target": "button",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  variant="primary"
/>
`;
